### ISSUE TITLE
GOTO file path

### ISSUE THREAD

**Issue by crabdancing** [nix-community/nixd#557]
**Is your feature request related to a problem? Please describe.**

It's awkward to navigate between related source files (e.g., source files that import other source files).

**Describe the solution you'd like**

LSP go-to on file path would ideally open that path.

**Describe alternatives you've considered**

I've tried to setup multi-LSP support and have `nil` do it, but it doesn't seem to work.

**Additional context**

In [nil](https://github.com/oxalica/nil), if you 'go to' a path, it will open that file path. E.g., 'go to `./module.nix`' will open `./module.nix`. After switching to `nixd`, I am loving a lot of the cross-file analysis, but I really miss being able to hop between source files.

**Comment by inclyc** [comment]
Did you mean "go to definition" for file paths?

AFAIK, Ctrl + Click works as a part of "documentLink" on vscode, I guess your editor may not provide such linking.

**Comment by crabdancing** [comment]
Yeah, I believe 'go to definition' is what it is. In Helix, it triggered when I hit `g d`, which is probably the exact same thing as Ctrl+Click in VSCode. My config for Helix (setup via home-manager) is as follows:

```
languages = {
      language-server.nixd = {
        command = "nixd";
        formatting = {
          command = ["alejandra"];
        };
        options.nixos = {
          expr = "(builtins.getFlake \"/etc/nixos\").nixosConfigurations.core.options";
        };
      };
      language = [
        {
          name = "nix";
          auto-format = true;
          language-servers = ["nixd" "nil"];
          formatter.command = "${pkgs.alejandra}/bin/alejandra";
        }
      ];
    };
```

**Comment by inclyc** [comment]
I see, seems to be reasonable to integrate it.

**Comment by crabdancing** [comment]
If it helps, I looked into it and figured out how to get a log of the Helix LSP thread's interactions with `nixd`. The output is [here](https://gist.github.com/crabdancing/42570a80040ab408e7e2209f8c5fe384).

**Comment by inclyc** [comment]
> If it helps, I looked into it and figured out how to get a log of the Helix LSP thread's interactions with `nixd`. The output is [here](https://gist.github.com/crabdancing/42570a80040ab408e7e2209f8c5fe384).
> 
> As a point of comparison, [here](https://gist.github.com/crabdancing/5c3a5d06f472caaaf1cf77970023057c) is Helix interacting with `nil` to open a file by path via go-to-definition.

Thanks for the detail, *ack*ed.

**Comment by crabdancing** [comment]
Thanks for that! I really appreciate it. ^w^

**Comment by Brisingr05** [comment]
I use Helix and `g d` works for me (unless I'm misunderstanding the issue):

https://github.com/user-attachments/assets/a17e4e4f-13b2-4cb1-8b4e-45b21bb306ed

Relevant snippet from HM config for Helix:

```nix
languages = {
  language = [
    {
      name = "nix";
      auto-format = true;
      language-servers = [ "nixd" ];
      formatter.command = "nixfmt";
    }
  ];
  language-server = {
    nixd = {
      command = "nixd";
      config = {
        nixpkgs.expr = "import (builtins.getFlake \"/home/brisingr05/nixos-config\").inputs.nixpkgs { }";
        formatting.command = [ "nixfmt" ]; # Currently doesn't work
        options.nixos.expr = "(builtins.getFlake \"/home/brisingr05/nixos-config\").nixosConfigurations.hpaio.options";
      };
    };
  }
}
```

**Comment by Brisingr05** [comment]
(Please let me know if I'm misunderstanding the issue):

In Helix, `g f` is "Goto files/URLs in selections" and `g d` is "Goto definition". In the video I press `g f` followed by `g d`:

https://github.com/user-attachments/assets/446be114-eab6-4961-89de-20961f7f2802

Relevant snippet from HM config for Helix:

```nix
languages = {
  language = [
    {
      name = "nix";
      auto-format = true;
      language-servers = [ "nixd" ];
      formatter.command = "nixfmt";
    }
  ];
  language-server = {
    nixd = {
      command = "nixd";
      config = {
        nixpkgs.expr = "import (builtins.getFlake \"/home/brisingr05/nixos-config\").inputs.nixpkgs { }";
        formatting.command = [ "nixfmt" ]; # Currently doesn't work
        options.nixos.expr = "(builtins.getFlake \"/home/brisingr05/nixos-config\").nixosConfigurations.hpaio.options";
      };
    };
  }
}
```

**Comment by crabdancing** [comment]
No, you're totally interpreting correctly, @Brisingr05. That is very strange. All I get when I try this is:

![image](https://github.com/user-attachments/assets/a1e7db28-a0aa-402b-841b-3dfe017294c1)

I doubt it's a bug in my Helix version. I've swapped between my fork of 24.05, stock 24.07, etc. I'm currently on nixd v2.3.0.

I'll try replicating your language-server as close as possible (obviously, `nixosConfigurations` target has to change) and see what happens.

**Comment by crabdancing** [comment]
Just tested it, and it also works in `clang` in the context of C projects.

`g d` in Helix takes you to definition for e.g. `#include <stdio.h>`. I think it's reasonable to say that this behavior is fairly well standardized, at least for static analysis of systems programming languages. :P
